apps/omg/lib/omg/typed_data_sign.ex

```diff
@@ -0,0 +1,98 @@
+# Copyright 2018 OmiseGO Pte Ltd
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+defmodule OMG.TypedDataSign do
+  @moduledoc """
+  Verifies typed structured data signatures (see: http://eips.ethereum.org/EIPS/eip-712)
+  """
+
+  alias OMG.Crypto
+  alias OMG.State.Transaction
+  alias OMG.Utxo
+
+  require Transaction
+  require Utxo
+
+  @domain_separator __MODULE__.Config.compute_domain_separator_from_config()
+  @transaction_encoded_type "Transaction(" <>
+                              "Input input0,Input input1,Input input2,Input input3," <>
+                              "Output output0,Output output1,Output output2,Output output3," <>
+                              "bytes32 metadata)"
+  @input_encoded_type "Input(uint256 blknum,uint256 txindex,uint256 oindex)"
+  @output_encoded_type "Output(address owner,address token,uint256 amount)"
+
+  @transaction_type_hash Crypto.hash(@transaction_encoded_type <> @input_encoded_type <> @output_encoded_type)
+  @input_type_hash Crypto.hash(@input_encoded_type)
+  @output_type_hash Crypto.hash(@output_encoded_type)
+
+  @empty_input_hash "1a5933eb0b3223b0500fbbe7039cab9badc006adda6cf3d337751412fd7a4b61" |> Base.decode16!(case: :lower)
```

Comment:
either explain in comment how to get this value, or just generate it using sth like `{0,0,0} |> some_hashing`.

Similar comment applies to other magic values

Reply:
Can't be directly computed because functions that computes them are defined in the same module. Commented.

Reply:
oh, if this is the reason here, then maybe rearranging the modules is the way to go? like with the domain separator?

Reply:
First I can't see nice way to rearrange - should I move very similar functionality because of this?
This is just an optimization and I think I could leave it as is now with a comment or compute this every time when needed.

Reply:
@pnowosie if this is part of `Signing` EIP, then put the module (and all subsequent modules) into `omg/lib/omg/sign/typed_data.ex` and this way, you'll have a separate namespace and you can further split this module into whatever makes these precomputed variables more understandable and readable.

Reply:
I see your point and I failed to express my issue here. I don't feel comfortable splitting TypedData module, because it feels correct to me to keep very similar functionality `hash_tx`, `hash_input` & `hash_output` together. All of them do the same thing, but for different input.

Reply:
I guess introducing sth like `OMG.TypedDataHash.Input.hash/1` or `OMG.TypedDataHash.Tools.hash_input/1` is the rearrangement that would help here.

I agree it would fragment code a bit, but I think it will pay off. Reducing the amount of confusion for someone who'll end up with a mismatched hash is important.

Reply:
Thanks for drawing my attention to this issue. I think I've found a nice way to rearrange modules while keeping similar things together.